### PR TITLE
[MICA-14067] saml2 4.0.2 rebase arterys

### PR DESCRIPTION
**Pull request by nsvarich**
This PR updates the Arterys saml2 fork with the latest upstream, by taking [upstream saml2 master 4.0.2 ](https://github.com/Clever/saml2/tree/0f40b833f55ca227702f864eb17947dd5b78c4be)and cherry-picking most of the existing Arterys fork changes, plus some additional unit test fixes. Cherry-picking was the simplest way as merging upstream had too many complicated conflicts.

### Arterys Changes Kept
- Attach saml namespace definition before stringifying the document and losing this information https://github.com/morpheus-med/saml2/commit/af47d22b62415b780964ea069e8bb259cc45615a
- Adds assert.user.name_id_format to the output https://github.com/morpheus-med/saml2/commit/06374e5b0f3758336ee626c3b9a05165933419f7, https://github.com/morpheus-med/saml2/commit/09f4521b128ca3e97fa8c57415c46a9c4422aacc
- Add phone numbers to assertion map https://github.com/morpheus-med/saml2/commit/ccf92380cd208012a8f2df9d86590267e74b8f6b
- VWR-458 Fixes to support 3rd party IDPs
https://github.com/morpheus-med/saml2/pull/3 
- VWR-458 Fix query parameter
https://github.com/morpheus-med/saml2/pull/4 
- VWR-458 NameID Fixes
https://github.com/morpheus-med/saml2/pull/5 
- MICA-14067 Updates xml-crypto to v5, to fix the AD FS cert validation error https://github.com/morpheus-med/saml2/pull/8

### Arterys Changes skipped
- Bug Fix ADFS 3.0
https://github.com/morpheus-med/saml2/compare/2019.01...morpheus-med:2021.06?expand=1 - This fix is replaced with an alternate simpler fix https://github.com/morpheus-med/saml2/pull/8
- Various minor build specific changes that are no longer relevant

### CODE REVIEW

**Review by tjugdev — APPROVED** [review]

File: test/saml2.coffee [hunk 1]
```diff
@@ -128,11 +132,11 @@ describe 'saml2', ->
...
         assert(
           has_attribute(
-            logout_service, 'Binding', 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect'),
+            logout_service, 'Binding', 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST'),
           "Expected to find an SingleLogoutService with redirect binding.")
 
         assert(
-          has_attribute logout_service, 'Location', 'https://sp.example.com/assert',
+          has_attribute logout_service, 'Location', LOGOUT_ENDPOINT,
           "Expected to find an SingleLogoutService with location 'htps://sp.example.com/assert'")
```
Comment (nsvarich) [inline test/saml2.coffee:140]: Change to logout

File: test/saml2.coffee [hunk 2]
```diff
@@ -128,19 +132,29 @@ describe 'saml2', ->
...
+        
+        assert _(entity_descriptor.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:metadata', 'AssertionConsumerService')).some((assertion) ->
+          _(assertion.attributes).some((attr) -> attr.name is 'Binding' and attr.value is 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST') and
+            _(assertion.attributes).some((attr) -> attr.name is 'Location' and attr.value is ASSERT_ENDPOINT))
+          , "Expected to find an AssertionConsumerService with POST binding and location 'https://sp.example.com/assert'"
+
+        assert _(entity_descriptor.getElementsByTagNameNS('urn:oasis:names:tc:SAML:2.0:metadata', 'SingleLogoutService')).some((assertion) ->
+          _(assertion.attributes).some((attr) -> attr.name is 'Binding' and attr.value is 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST') and
+            _(assertion.attributes).some((attr) -> attr.name is 'Location' and attr.value is LOGOUT_ENDPOINT))
+          , "Expected to find a SingleLogoutService with redirect binding and location 'https://sp.example.com/logout'"
```
Comment (hinshaw-arterys) [inline test/saml2.coffee:157]: Says "with redirect binding", but looks like a POST binding 2 lines above?  (bindings:HTTP-POST)

**Review by hinshaw-arterys — APPROVED** [review]

File: test/data/good_assertion.xml [hunk 3]
```diff
@@ -49,9 +48,9 @@
...
     </Attribute>
   </AttributeStatement>
-  <AuthnStatement AuthnInstant="2014-03-12T21:35:05.354Z" SessionIndex="_3">
+  <AuthnStatement AuthnInstant="2014-03-12T21:35:05.354Z" SessionIndex="_3" SessionNotOnOrAfter="2014-03-13T22:35:05.387Z">
     <AuthnContext>
       <AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</AuthnContextClassRef>
     </AuthnContext>
   </AuthnStatement>
-</Assertion>
+</Assertion>
```
Comment (hinshaw-arterys) [inline test/data/good_assertion.xml:56]: Missing final newline

File: test/data/good_response_twice_signed.xml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                <AttributeValue>Test</AttributeValue>
+            </Attribute>
+        </AttributeStatement>
+        <AuthnStatement AuthnInstant="2014-03-12T21:35:05.354Z" SessionIndex="_3">
+            <AuthnContext>
+                <AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</AuthnContextClassRef>
+            </AuthnContext>
+        </AuthnStatement>
+    </Assertion>
+</samlp:Response>
```
Comment (hinshaw-arterys) [inline test/data/good_response_twice_signed.xml:38]: missing final newline

File: test/data/good_response_twice_signed_dsig_ns_at_top.xml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        <AttributeValue>Test</AttributeValue>
+      </Attribute>
+    </AttributeStatement>
+    <AuthnStatement AuthnInstant="2014-03-12T21:35:05.354Z" SessionIndex="_3">
+      <AuthnContext>
+        <AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</AuthnContextClassRef>
+      </AuthnContext>
+    </AuthnStatement>
+  </Assertion>
+</samlp:Response>
```
Comment (hinshaw-arterys) [inline test/data/good_response_twice_signed_dsig_ns_at_top.xml:39]: missing final newline

File: test/data/response_external_signed_assertion.xml [hunk 6]
```diff
@@ -0,0 +1 @@
...bz48Q2Fub25pY2FsaXphdGlvbk1ldGhvZCBBbGdvcml0aG09Imh0dHA6Ly93d3cudzMub3JnLzIwMDEvMTAveG1sLWV4Yy1jMTRuIyIvPjxTaWduYXR1cmVNZXRob2QgQWxnb3JpdGhtPSJodHRwOi8vd3d3LnczLm9yZy8yMDAxLzA0L3htbGRzaWctbW9yZSNyc2Etc2hhMjU2Ii8+PFJlZmVyZW5jZSBVUkk9IiNfMiI+PFRyYW5zZm9ybXM+PFRyYW5zZm9ybSBBbGdvcml0aG09Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvMDkveG1sZHNpZyNlbnZlbG9wZWQtc2lnbmF0dXJlIi8+PFRyYW5zZm9ybSBBbGdvcml0aG09Imh0dHA6Ly93d3cudzMub3JnLzIwMDEvMTAveG1sLWV4Yy1jMTRuIyIvPjwvVHJhbnNmb3Jtcz48RGlnZXN0TWV0aG9kIEFsZ29yaXRobT0iaHR0cDovL3d3dy53My5vcmcvMjAwMC8wOS94bWxkc2lnI3NoYTEiLz48RGlnZXN0VmFsdWU+S1dPVllha0JVQXgzbFhqQm5BeExiRHJWRzhRPTwvRGlnZXN0VmFsdWU+PC9SZWZlcmVuY2U+PC9TaWduZWRJbmZvPjxTaWduYXR1cmVWYWx1ZT5qQjBHbW85anVmZk9VVXY5bWIvaU1iSXA5YjFxZTkxL0RtMUZCcytTb0tMTzh1KzZ4T0xNTHArQ1h2bmxMK3VpVFlnVE1HYmRGWjZlNnpFSG96TFNCeWpOWjQ3aWVHVG9ocGh0c2VqQitZMTdDNUZjclQxS05ZVWtPMkNZd0l6cSt4TVJwK0MrQXIwQTJoRzRiMktxT0drc2JrZWp0TDY3bVNiNG9KaExKTUZ4Mk5FaU5VN1E2eTlGeHhJZW5IS0hXUFgzeEVxdHFVU3RvcXhRY2p3aWpYTm9IMUdKZWJUWEsvQW9NZG10bTMzS0pON1ova1lkWlFZdndGRmg2dzZHZTdVZVB4OVVXQmtUZUNSLytNb25RTG1lUzQwbDBHb3ZRU3BiYkdZd0RUbWlXYmZwVFlYdTFGRGN1cVMrTFkxM1RwTFpJai9BVHRHajM3Qm94Uk4zVXc9PTwvU2lnbmF0dXJlVmFsdWU+PC9TaWduYXR1cmU+PC9zYW1scDpSZXNwb25zZT4=
```
Comment (hinshaw-arterys) [inline test/data/response_external_signed_assertion.xml:1]: missing final newline

File: test/data/response_notbefore_future_encrypted.xml [hunk 7]
```diff
@@ -0,0 +1 @@
...M040L0RKZnVuRU9zN2tvVzhGR00zVWdnZ3h1N1VxY3lRNVphLzdVVmxoMmhtUFpsc09TNFlSNjh2VHFmaVYwQldsdmNoaE5ORTFxL3RYTWMxeHNhbGJVOS85UVcxcGxYTzMyb0F3ajVnSG0rcERSSUluakRKMDkxTFEzdnNweGVBbWJwOWJBYisyRC9DSXJaU21MWVhZMTY4aGFpVFFPYURHZXdtOElncm15UnhEWHkwWDFOU0FhL1U5Szd4cjZWNVhoaVVBVnVyVGN3Y1pnTVNOS29GMllwMlo5ajE4d0hpU09nd1Rna25UN0Z4K3BPMVdRMVBPVFJkN2JGa3RlMHR2a1BQaCttSk40cmN2N2ZpT3ljaEVnckdYNUo4YVp5ZUk5L2F6Z0RCVzdFdE1IRWNIakkyTFR0bEdXaG8wVjhBbzVFd09zd3lkRU5iTUFFNVBoWFhNc29kUXdadm1LZEt6U1AzWkR1VEhSUWVHcTdwc2NQc20xWERtZWxZOFFEeC84RUFpNzAwNy9leXA0Z1BKZVBTVlF3bHJGamFzS20rS1dQc1NqWG5CTGZ1TzVjOFZ6SDFhMDk5dlBxYTVZQzFQUUdPd3B4ZENxUG5HRmxuK2pTa05oaTNQcTZKNnZiTjVmbXBQS3VuWks3Y2E5b3VXNGlYRTBoUGhPY3l4bWZFeVV4R3NPQWtLSlBlZzdSVjRMUnpDNnB0cElDc0NXNU9PdW5nblhVSElWVnlRN0hXbDhBTStHclI2VjBidjBEcjh3U29IMmpjakRMclYxRHhtZktjd1kxV2JVNHZEWlBKYjFBZy85c2VJNEdwTmIyOFdmY3I5QUNhdlVsKzJEVjR2VXFKckRQRTFwRzFUVm1uUHhzQnNIbURiTThGL0FKUXRjRkNNYjRQcDdXbytrU3VWNVVibDlrR0VDMitGTkVTWlJsdlUxTVNNOFVBeGhkOHp0Tk1xcitVMTFDM2JlcWo2WjFpT1dDN21VU1dJMjByT2crTi9CY0pVUkZ4Mmo5QVpnS2J6dnBHcG4rSURSUXB2Sy9Saz08L3hlbmM6Q2lwaGVyVmFsdWU+DQogICA8L3hlbmM6Q2lwaGVyRGF0YT4NCjwveGVuYzpFbmNyeXB0ZWREYXRhPjwvc2FtbDpFbmNyeXB0ZWRBc3NlcnRpb24+PC9zYW1scDpSZXNwb25zZT4=
```
Comment (hinshaw-arterys) [inline test/data/response_notbefore_future_encrypted.xml:1]: missing final newline

File: test/data/response_notbefore_future_encrypted_then_signed.xml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+   </xenc:CipherData>
+</xenc:EncryptedData></saml:EncryptedAssertion></samlp:Response>
```
Comment (hinshaw-arterys) [inline test/data/response_notbefore_future_encrypted_then_signed.xml:15]: missing final newline

File: test/data/response_notbefore_future_signed.xml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                <AttributeValue>Test</AttributeValue>
+            </Attribute>
+        </AttributeStatement>
+        <AuthnStatement AuthnInstant="2014-03-12T21:35:05.354Z" SessionIndex="_3">
+            <AuthnContext>
+                <AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport</AuthnContextClassRef>
+            </AuthnContext>
+        </AuthnStatement>
+    </Assertion>
+</samlp:Response>
```
Comment (hinshaw-arterys) [inline test/data/response_notbefore_future_signed.xml:34]: missing final newline

File: test/data/response_notbefore_future_signed_then_encrypted.xml [hunk 10]
```diff
@@ -0,0 +1,10 @@
...
+   </xenc:CipherData>
+</xenc:EncryptedData></EncryptedAssertion></samlp:Response>
```
Comment (hinshaw-arterys) [inline test/data/response_notbefore_future_signed_then_encrypted.xml:10]: missing final newline

**Review by hinshaw-arterys — CHANGES_REQUESTED** [review]
Several files missing final newline.  (Not a huge deal, but I've seen that cause hard-to-solve weirdness before.)

File: test/data/good_assertion.xml (hunk 3 above)
Comment (hinshaw-arterys) [reply test/data/good_assertion.xml:56]: Nick found that these missing newlines are coming from upstream and we haven't touched them.  So let's leave them as-is.

**Pull request merged**